include/boost/math/special_functions/prime_sieve.hpp

```diff
@@ -123,44 +127,57 @@ void mask_sieve(Z lower_bound, Z upper_bound, Container &c)
 }
 } // End namespace detail
 
-template<typename Z, class OutputIterator>
-auto prime_sieve(Z upper_bound, OutputIterator output) -> decltype(output)
+#if __has_include(<execution>)
+template<class ExecutionPolicy, typename Z, class OutputIterator>
+auto prime_sieve(ExecutionPolicy&& policy, Z upper_bound, OutputIterator output) -> decltype(output)
 {
     static_assert(std::is_integral<Z>::value, "No primes for floating point types");
     BOOST_ASSERT_MSG(upper_bound + 1 < std::numeric_limits<Z>::max(), "Type Overflow");
 
     std::vector<Z> primes;
     primes.reserve(upper_bound / std::log(upper_bound));
 
-    if (upper_bound <= 104729)
+    if (upper_bound <= 8192)
     {
-        boost::math::detail::prime_table(upper_bound, primes);
+        boost::math::detail::linear_sieve(upper_bound, primes);
     }
 
     else
     {
-        std::vector<Z> small_primes;
-        small_primes.reserve(1000);
-
-        // Spilt into two vectors and merge after joined to avoid data races
-        std::thread t1([upper_bound, &small_primes]{boost::math::detail::prime_table(static_cast<Z>(104729), small_primes);});
-        std::thread t2([upper_bound, &primes]{boost::math::detail::mask_sieve(static_cast<Z>(104729), upper_bound, primes);});
+        if constexpr (std::is_same_v<decltype(policy), std::execution::sequenced_policy>)
+        {
+            boost::math::detail::mask_sieve(static_cast<Z>(2), upper_bound, primes);
+        }
 
-        t1.join();
-        t2.join();
-        primes.insert(primes.begin(), small_primes.begin(), small_primes.end());
+        else
+        {
+            std::vector<Z> small_primes;
+            small_primes.reserve(1000);
+
+            // Split into two vectors and merge after joined to avoid data races
```

Comment:
Can you split this into as many vectors as threads?

Reply:
I can look into that today. I do not see why not.